feat: execution for fragment arguments syntax

JoviDeCroock opened the pull request:
This implements execution of Fragment Arguments, and more specifically collecting-fields and replacing the variables in a fragment-definition with values coming from the fragment-spread, as described by the spec changes in https://github.com/graphql/graphql-spec/pull/1010.

This is a 2024 update of https://github.com/graphql/graphql-js/pull/3835 where all discussions have been carried over in code comments, they are highlighted already as review comments here as well.

This PR splits out the execution part so we can look at that in detail, I will pause here as the fragment-keying remarks will most likely have quite an impact on the validation changes.

JoviDeCroock committed:
implement execution for fragment arguments syntax

Co-authored-by: mjmahone <mahoney.mattj@gmail.com>

JoviDeCroock commented on src/utilities/substituteFragmentArguments.ts:
```diff
@@ -0,0 +1,79 @@
+      // replacing an unset fragment argument with an operation
+      // variable value. Fragment arguments must always have LOCAL scope.
+      //
+      // To remove this hack, we need to either:
```
The issue here presents itself in that fragments can define arguments like

```
fragment x($id: ID!) on Y { todo(id: $id) { text } }
```

Which can be passed in through fragment spreads

```
...x(id: '2')
```

now what happens when we spread `x` and there is an `$id` present on the document itself, if we aren't using `__UNSET` it would leverage that one which we don't want. There might be a better solution in carrying `variables` around as we visit the `FragmentSpread` and the accompanying `FragmentDefinition`

Relevant conversation on this: https://github.com/graphql/graphql-js/pull/3835#discussion_r1101008869

We could so something along the lines of 

```js
  return visit(def.selectionSet, {
    Argument: {
      leave: (node) => {
        // if null remove the argument
        if (value === UNSET) {
          // Remove argument completely
        } else if (arg === LIST) {
          // Iterate over list and replace any UNSET with a NULL kind
        } else if (arg === INPUT_OBJECT) {
          // not sure whether we omit the key or replace with null as well
        }
      }
    },
    Variable(node) {
      const value = argumentValues.get(node.name.value);
      if (value?.kind === Kind.VARIABLE && value.name.value === UNSET) {
        return null
      }
      return value;
    },
  });
```

But that could get complex with values like `[$x, 'y']` as we most likely want to keep y in position 1. We implicitly rely on `__UNSET` not being a defined variable on the operation so that at replacement time this gets seen as `null`.

JoviDeCroock replied:
An alternative I have been playing around with is to basically create a new `argumentsContext` when we go from a FragmentSpread into Fields within the collectFields, we would treat the `FragmentDefinition` variables similar to an operation where we calculate a scope and pass it into the `FieldInfo` so we can sue that in execute. I do think that the over-arching refactor could be quite big though.

We could deem `__UNSET` good enough for now or prefix it with the variable-name so it becomes `${var.name}__UNSET` as a v1 to not meddle too much with the current internals

Tried an approach like the following

```js
          const rawVariables = fragment.variableDefinitions?.reduce((acc, variableDefinition) => {
            const varName = variableDefinition.variable.name.value;
            const argumentForVariable = selection.arguments?.find(arg => arg.name.value === varName);
  
            if (argumentForVariable) {
              return { ...acc, [varName]: valueFromASTUntyped(argumentForVariable.value, acc) }
            }
  
            if (variableDefinition.defaultValue) {
              return { ...acc, [variableDefinition.variable.name.value]: valueFromASTUntyped(variableDefinition.defaultValue, acc) }
            }
  
            return acc;
          }, variableValues)
```

However then we run into issues where we want to take the type in account which `valueFromAST` does not seem to support (hence I used untyped for now)

JoviDeCroock replied:
Another attempt https://github.com/JoviDeCroock/graphql-js/compare/fragment-args-execution-2024...JoviDeCroock:graphql-js:collect-fields-alternative?expand=1

JoviDeCroock replied:
Relevant https://benjie.dev/rfcs/1056

mjmahone replied:
> We need to introduce a concept of shadowing:
> - when a fragment defines a variable that is in the parent scope but not given
>    in the fragment-spread we need to look at this variable as undefined and check
>    whether the definition has a defaultValue, if not remove it from the variableValues.
> - when a fragment does not define a variable we need to copy it over from the parent
>    scope as that variable can still get used in spreads later on in the selectionSet.
> - when a value is passed in through the fragment-spread we need to copy over the key-value
>    into our variable-values.

I'm not sure this is *quite* correct. Specifically, "remove it from the variableValues": are you saying "remove it from the variableValues *to use for this specific fragment*" or "remove from variableValues for all future execution"? I can't quite tell yet.

Those two imply very different semantic "scopes". What we want to avoid is adding in some form of "contextual" (lexical?) scope. Operation variables currently have "dynamic" scope, i.e. the definition of the variable (including its type and default value) depends on *which* operation you're executing. We absolutely don't want a *second* form of dynamic scoping, where child fragment spreads "inherit" their parent fragments' argument definitions. Nor do we want a parent fragment to "wipe out" an operations' variable definition, *except* within the context of *that one fragment*.

JoviDeCroock replied:
It removes it from the new scope i.e. `variableValues` is altered when we are collecting field for said fragment-spread and subsequent nested fragment-spreads in the corresponding definition

mjmahone replied:
> variableValues is altered when we are collecting field for said fragment-spread and subsequent nested fragment-spreads in the corresponding definition

I don't think that's what is wanted.

```
query A($x: Int = 3) {
  ...B
}

fragment B($x: String) on Query {
  echo(value: $x)
  ...C
}

fragment C on Query {
  howMany(count: $x)
}
```

`howMany(count: $x)` ought to resolve to `howMany(count: 3)` not `howMany(count: null)`

JoviDeCroock replied:
That is exactly what is happening, the tests are all passing. When a variable isn't redefined on the new definition it implicitly transfers to the child-scope in shadowing. as laid out here https://github.com/JoviDeCroock/graphql-js/pull/3

EDIT: added the above test in https://github.com/JoviDeCroock/graphql-js/pull/3/commits/588428cc4fa2497e0aed0a41d0c4714dc7135292 which works great

mjmahone replied:
Ah sorry I was misunderstanding which scope `variableValues` implied.

JoviDeCroock commented on src/execution/collectFields.ts:
```diff
@@ -173,7 +174,7 @@ function collectFieldsImpl(
       }
       case Kind.FRAGMENT_SPREAD: {
-        const fragName = selection.name.value;
+        const fragmentName = selection.name.value;
```
This was changed from the original, before we used to generate a key for each FragmentSpread to see whether it was the same as others in the selection-set however this isn't needed as discussed in https://github.com/graphql/graphql-js/pull/3835#discussion_r1101825832

mjmahone commented:
Yeah I realized (after I started) that we do need to preserve a two-layer “stack” of variables, a la the refactored context you’re describing. Substitution doesn’t cut it: I can’t remember all the details for WHY substitution doesn’t work (sounds like you’re learning them though!).

Yes it would be a pretty large refactor, but I think that refactor is essentially required to make fragment arguments good.

JoviDeCroock commented:
@mjmahone the refactor is largely done in https://github.com/JoviDeCroock/graphql-js/compare/fragment-args-execution-2024...JoviDeCroock:graphql-js:collect-fields-alternative?expand=1

yaacovCR commented:
Overall I wonder if it would be nice to have some common GraphQLArgument/GraphQLSignature type similar to https://github.com/JoviDeCroock/graphql-js/pull/7 that could unify some of the above logic.

yaacovCR commented:
If it helps at all, I have rebased this branch on current main: https://github.com/yaacovCR/graphql-executor/tree/fragment-args-execution-2024-rebased

yaacovCR commented:
Here is a commit that implements some parallel changes to what I suggested in the TypeInfo branch with precompiling the fragment variable signatures....

https://github.com/yaacovCR/graphql-executor/commit/e5744fc79ee560eaa55796faeef8dd5261dd912f

Also includes greater code reuse of getArgumentValues, but presently requires some JavaScript specific changes around non-provided variables, previously they did not appear at all in the args array, now they need to be present but undefined, so the code path for supplying the default field value takes precedence over the operation default value

yaacovCR commented on src/execution/collectFields.ts:
```diff
@@ -129,21 +133,24 @@ function collectFieldsImpl(
     switch (selection.kind) {
       case Kind.FIELD: {
-        if (!shouldIncludeNode(variableValues, selection)) {
+        const vars = localVariableValues ?? variableValues;
```
why not both? as `getDirectiveValues` ultimately calls `getArgumentValues`, might it not be necessary to track both the global and local variableValues through? This should be true whenever/wherever directives are used.

JoviDeCroock replied:
Hmm, I assume this comment is misplaced then, you mean that https://github.com/JoviDeCroock/graphql-js/pull/2/files#diff-1d705c6a4c73cd3ce46190029e75abd3015b49de3ec250357dcf9b44a35cb7d0R931 should become `{ ...exeContext.variableValues, ...fragmentVariableValues }`?

yaacovCR replied:
Maybe that should be looked at as well... But the problem I am referencing here --- somewhat circuitously --- is that `shouldIncludeNode` passes the variableValues to `getDirectiveValues` which then calls `getArgumentValues`. `getArgumentValues` requires both the local and the global variableValues, but it isn't receiving them.

Overall, I wonder if we could consider abandoning the coalescence of globalVariableValues into the localVariables; instead I think it would be better to simply retain references to both the global and local variables, and have the getArgumentValues function perform whatever analysis it needs. I believe also the local variable definitions/signatures would also have to be passed to getArgumentVariables and I would suggest a single additional argument like:

```ts
interface ExperimentalFragmentVariables {
   signatures: ObjMap<GraphQLFragmentVariableSignatures>;   // the defs with the types validated and the defaults calculated
   variableValues: ObjMap<unknown>;
} 
```

The idea of preserving the signatures/definitions in an interface is similar to the approach by @leebyron with respect to default value validation (see mega-PR https://github.com/graphql/graphql-js/pull/3049  and specifically https://github.com/graphql/graphql-js/pull/3077/files#diff-777202fe667b9e656f40631c341c7c0776895a75eff7004578bf9b7a97e095abR29-R37) and may be the direction we will be heading in anyway whenever that long awaited work is merged.

yaacovCR replied:
[This PR](https://github.com/JoviDeCroock/graphql-js/pull/9) demonstrates what I believe is a test failure. Let me know what you think.

Your first comment could have a potential fix to that, but the link is not working for me right now, I think it pointed to here: https://github.com/JoviDeCroock/graphql-js/pull/2/files#diff-1d705c6a4c73cd3ce46190029e75abd3015b49de3ec250357dcf9b44a35cb7d0L726, which would be only for the stream directive, although you could do the same thing for skip/include, but I think that would also not work, as to deal with the shadowing case, `getArgumentValues` has to know that there was a fragmentVariable defined and not provided, so `getDirectiveValues` would have to know that there are actually fragmentVariables just like the call to `getArgumentValues` passes fragmentVariables in.

yaacovCR commented on src/execution/collectFields.ts:
```diff
@@ -34,15 +34,17 @@ export interface DeferUsage {
   operation: OperationDefinitionNode;
   runtimeType: GraphQLObjectType;
   visitedFragmentNames: Set<string>;
+  localVariableValues: { [variable: string]: unknown } | undefined;
```
Is it safe to store the `localVariableValues` in the context? Might this not get rewritten by nested fragments?

JoviDeCroock replied:
The goal was to not rewrite it which is successful due to recalculating at every intersection, this also is meant to support using the variables in a fragment-definition in a future spread.

yaacovCR replied:
What I'm worried about is a document like the following:

```graphql
        query {
          ...A(value: "A")
        }
        fragment A($value: String!) on Query {
          ...B(value: "B")
          someFieldA(input: $value)
        }
        fragment B($value: String!) on Query {
          someFieldB(input: $value)
        }
}

```
When I coalesced the parameters to CollectFieldsImpl into a `context`, the intent was to do so because these variables were essentially readonly (e.g. schema) or accumulators (e.g. newDeferUsages) and so there could be no conflicts.

JoviDeCroock replied:
This should all be safe because of how we reset this context during collectFields, every spread entry will either take this to undefined _or_ overwrite it with the defined variables

yaacovCR replied:
I see, the test passes, I didn't catch/realize that that even though you are resetting the `localVariableValues` on the context for the fragment, the later fields will pick up the value from the `localVariableValues` that were already destructured here: https://github.com/JoviDeCroock/graphql-js/blob/2d354bcd7d567c71cd2c535a9dde25fd5d260ad4/src/execution/collectFields.ts#L159

Because you are using the context to basically pass a different value to the inner `collectFieldsImpl`, I would still suggest refactoring as an additional argument, but you are correct that there is no bug in this case.

yaacovCR replied:
if you would like here is a PR with those tests to show that they pass: https://github.com/JoviDeCroock/graphql-js/pull/8

yaacovCR commented:
> but presently requires some JavaScript specific changes around non-provided variables, previously they did not appear at all in the args array, now they need to be present but undefined

I was able to fix this on my scratch brainstorming branch: https://github.com/yaacovCR/graphql-executor/tree/execution-suggestions

yaacovCR commented:
> TODO: rebase last 1 commit on fragments args syntax

branch https://github.com/yaacovCR/graphql-executor/tree/execution-suggestions has that commit rebased on main and then one additional commit showing you one potential set of changes that would be responsive to the feedback i am trying to give.

Basically, in that branch:

1. we reuse getArgumentValues for field arguments and fragment arguments
2. we do not coalesce fragmentVariablesValues and regular variableValues into a single map, but just pass around both to the relevant functions
3. we have an `experimental` prefix for the function that changes signature (`experimentalGetArgumentValues()`), everything else just gets optional arguments
4. we do not store fragmentVariables in the context
5. we have a separate type GraphQLVariableSignature that we use whenever we encounter a variable ~~(this last item actually fixes a trivial bug with one-of)~~

yaacovCR committed:
add directive test (#9)

* add directive test

* add failing test

add additional nested fragment test (#8)

Correct test and lint stuff

suggestions for execution (#11)

* introduce internal getVariableSignature utility

now extracted also to graphql-js PR, see https://github.com/graphql/graphql-js/pull/4175

* execution suggestions

fixes execution to always use fragment variable when has the same name as an operation variable

previously, we were allowing an operation variable to be used if the fragment variable was not provided, and the field had no default. Now, we still use the fragment variable, and so the value is null.

this now correct logic allows us to significantly reduce the diff from main

adds additional test

JoviDeCroock commented on src/execution/__tests__/variables-test.ts:
```diff
@@ -1434,5 +1461,33 @@ describe('Execute: Handles inputs', () => {
+      });
+    });
+
+    it('when a nullable argument to a directive with a field default is not provided and shadowed by an operation variable', () => {
```
@yaacovCR I think this test is incorrect because we aren't providing `$value` and the directive skip has a non-nullable if. The other tests however all pass, and this one correctly fails 😅

yaacovCR replied:
The point of the test is that we are providing a default for our custom skip directive and so that default should be grabbed just as it is in the tests above.

https://github.com/yaacovCR/graphql-executor/blob/1ea79ef6c0a29d521c51252106ac4af2ba55a7b0/src/execution/__tests__/variables-test.ts#L177-L178

In the light of day, however, it appears that `collectFields()` does not call `schema.getDirective('skip')` to find our custom skip directive. I am not 100% sure that is the correct behavior, but it seems besides the point right now.[^1]

In any case, if you manually change the reference in `collectFields()` to `schema.getDirective('skip')` instead of `GraphQLSkipDirective`, it should pick up our default, and it doesn't, see https://github.com/JoviDeCroock/graphql-js/pull/2#discussion_r1730094734

I will change the test, though, because it is kind of worthless in the end considering how collectFields doesn't pick up that default that we added!

[^1]: Generally speaking, It is valid to have custom runtime skip/include/defer/stream directives, although within the graphql-js explicitly any custom runtime directives would presumably have a custom version of `collectFields`, so it seems kind of harmless, except that it prevents us from adding a default value to `@skip(if:)` like in our test. Of course, `@skip(if:)` has a default value of `true` because when the argument is omitted, it will just skip by design, so it doesn't need a default. I only added the default to check how fragment variables will work. So I don't think it's really worth it to "fix" `collectFields()`. Calling  `schema.getDirective('skip')` is just one extra function call that isn't really necessary.

yaacovCR committed:
move getVariableSignature to execution

as it cannot be used by validation, which must collect all errors rather than fail with invalid type for signature
